### PR TITLE
add vhost-user-vsock device emulation

### PR DESCRIPTION
**Pull request by stefano-garzarella**
This PR is based on https://github.com/rust-vmm/vhost-device/pull/7.

@harshanavkis implemented a _vhost-user-vsock_ device for a [GSoC 2021 project](https://summerofcode.withgoogle.com/archive/2021/projects/6126117680840704) mentored by me and @fidencio.

His PR was in a great state, I just rebased and completed the following items to be able to merge this project that could be very useful:
- updated some dependencies (e.g. vm-memory, etc.)
- replaced the deprecated `App::from_yaml()`
- fixed some new warnings from `clippy`
- addressed @stsquad comments (don't panic when `CONNECT 1234` is wrong, use [virtio-vsock 0.1.0](https://crates.io/crates/virtio-vsock) crate from @lauralt )
- avoided sending credit updates when the credit hasn't changed
- added `SPDX-License-Identifier` in each vsock file

### CODE REVIEW

**Comment by stefano-garzarella** [comment]
v2:
- fixed more clippy warnings (I run the nightly clippy to be sure I cover also the new ones)
- fixed commit title length issue

TODO:
- [ ] extend test coverage (this PR drops the coverage by 20%, would be nice to extend the tests. I'll try to do my best in the next days, but if someone wants to help, it will be great)

**Comment by stefano-garzarella** [comment]
v3:
- fixed another commit length issue
- reverted a change that I added by mistake in v2

TODO:
- [ ]  extend test coverage (this PR drops the coverage by 20%, would be nice to extend the tests. I'll try to do my best in the next days, but if someone wants to help, it will be great)

**Comment by stefano-garzarella** [comment]
v4:
- added more test to reach ~70% of coverage (it was ~55% before the new tests).
- removed some unwrap() and fixed little issues discovered while writing tests

I hope the coverage is not that bad, and we can merge this. In the future we can try to increase it, adding more moks.

File: vsock/src/thread_backend.rs
Comment (stsquad) [inline vsock/src/thread_backend.rs:25]: Why not convert now? Can we have multiple threads now?

File: vsock/src/main.rs
Comment (stsquad) [inline vsock/src/main.rs:42]: Should we do something here instead of just notify the user? Can we use error!() here for reporting?

File: vsock/src/main.rs
Comment (stsquad) [inline vsock/src/main.rs:57]: Use info! and warn! macros for reporting so we can follow the same standard reporting as the other daemons.

File: vsock/src/main.rs
Comment (stsquad) [inline vsock/src/main.rs:75]: I'd argue this is superflous verbiage

File: vsock/src/vhu_vsock_thread.rs
Comment (stsquad) [inline vsock/src/vhu_vsock_thread.rs:386]: superfluous commented out code?

File: vsock/src/vhu_vsock_thread.rs
Comment (stsquad) [inline vsock/src/vhu_vsock_thread.rs:481]: please remove commented out code

**Review by stsquad — CHANGES_REQUESTED** [review]
I've tested the daemon on my setup and everything looks pretty good to me. I have added a few comments about removing commented out code and cleaning up the println -> info!,warn! and debug! macros. I think the remaining TODOs and further increasing the coverage can be done once this is merged.

File: vsock/src/main.rs
Comment (stefano-garzarella) [reply vsock/src/main.rs:42]: I guess it was here just for debug purpose, so I would remove it or just add an assert!().
VhostUserDaemon takes this information from `VhostUserVsockBackend::queues_per_thread`.
In this case `VhostUserVsockBackend` setup a single thread and a single element in `queues_per_thread`, so this check should be always false.

File: vsock/src/main.rs
Comment (stefano-garzarella) [reply vsock/src/main.rs:57]: ack

File: vsock/src/main.rs
Comment (stefano-garzarella) [reply vsock/src/main.rs:75]: Agree, using `unwrap`

File: vsock/src/main.rs
Comment (stefano-garzarella) [reply vsock/src/main.rs:42]: Or maybe I can add a test in main.c to check these things.

File: vsock/src/vhu_vsock_thread.rs
Comment (stefano-garzarella) [reply vsock/src/vhu_vsock_thread.rs:481]: ack

File: vsock/src/thread_backend.rs
Comment (stefano-garzarella) [reply vsock/src/thread_backend.rs:25]: Nope, we have a single thread and I think that was an old TODO since `VsockThreadBackend` is supposed to be per-thread (IIUC).
Anyway, if we will move to multi-thread, I think we need to handle also other maps, so I would remove that one if you and  @harshanavkis agree.

Comment (stsquad) [reply vsock/src/thread_backend.rs:25]: fine by me

**Comment by stefano-garzarella** [comment]
v5:
- replaced all `println!()` with `dbg!()`, `info!()`, or `error!()`
- initialized `env_logger` in main()
- removed old comments, useless code, and code commented
- added test in main.rs to check that `VhostUserDaemon`'s workers and `VhostUserVsockBackend`'s threads are the same (they should be always the same if the code is correct)

**Comment by stefano-garzarella** [comment]
@stsquad thanks for testing and reviews! I just addressed all your comments and pushed the new version

**Review by stsquad — APPROVED** [review]

**Comment by stefano-garzarella** [comment]
> I will start looking into this pull request in the coming days. Given its sheer size I expect the review process to take several days. I will clearly let you know when I am done.

@mathieupoirier thank you very much! Yep, take your time. I have taken up Harsha's work, so some part may be unknown to me too ;-)

File: vsock/src/vhu_vsock.rs
Comment (vireshk) [inline vsock/src/vhu_vsock.rs:214]: Why __u64 here ?

Since we need to set this only once and then just send it in Little endian format, we can make it vm_memory::Le64 instead and set it as:

guest_cid = From::from(u64-value);

This is what's done in GPIO crate.

File: vsock/src/thread_backend.rs
Comment (vireshk) [inline vsock/src/thread_backend.rs:3]: still required ?

File: vsock/src/thread_backend.rs
Comment (vireshk) [inline vsock/src/thread_backend.rs:25]: pub(crate) should be enough ? I think that's true for most of the stuff here as this is a binary crate and most of the stuff shouldn't be visible out of this crate. I do see some violation of this in existing crates, will add an issue for that.

File: vsock/src/thread_backend.rs
Comment (vireshk) [inline vsock/src/thread_backend.rs:59]: Single line should be enough for this comment.

File: vsock/src/thread_backend.rs
Comment (vireshk) [inline vsock/src/thread_backend.rs:75]: What about:

        let key = match self.backend_rxq.pop_front().map_err(|_| Error::EmptyBackendRxQ)?;

File: vsock/src/thread_backend.rs
Comment (vireshk) [inline vsock/src/thread_backend.rs:112]: drop this and use if/else instead ?

**Review by vireshk — CHANGES_REQUESTED** [review]
About code coverage, I looked at kcov_output/ to see what all is missed and there are things we can easily cover I think. For example main.rs::start_backend_server() and a lot of other code in vhu_vsock_thread.rs which is not covered as of now in tests.

File: vsock/src/vhu_vsock.rs
Comment (vireshk) [inline vsock/src/vhu_vsock.rs:268]: Instead use std::io::Result directly.

    ) -> std::io::Result<()> {

For I2c/gpio, I have aliased that as IoResult (in a new pull request).

same applies to handle_event and elsewhere.

**Review by vireshk — CHANGES_REQUESTED** [review]

File: vsock/src/main.rs
Comment (stefano-garzarella) [reply vsock/src/main.rs:1]: I'm not the original author of this code, so I just added the SPDX-License-Identifier following what Harsha specified in his Cargo.toml.

I'm open to change on Apache-2.0 only, but we need an approval from @harshanavkis.

File: vsock/src/main.rs
Comment (stefano-garzarella) [reply vsock/src/main.rs:70]: This was by Harsha, I'm open to move VsockConfig and VsockArgs in main, if you prefer.

File: vsock/src/main.rs
Comment (stefano-garzarella) [reply vsock/src/main.rs:26]: I'll try to fix.

File: vsock/src/main.rs
Comment (stefano-garzarella) [reply vsock/src/main.rs:33]: I think Harsha followed what we did in QEMU where we have `vhost-user-blk`, `vhost-user-gpu`, `vhost-user-scsi`, `vhost-user-input`, etc. Honestly I would have done the same, at the most we could use `vhost-user-X-device`, though, I don't think _daemon_ adds much.

File: vsock/src/thread_backend.rs
Comment (stefano-garzarella) [reply vsock/src/thread_backend.rs:3]: I don't know, I'll check!

File: vsock/src/thread_backend.rs
Comment (stefano-garzarella) [reply vsock/src/thread_backend.rs:25]: okay, I'll change

File: vsock/src/thread_backend.rs
Comment (stefano-garzarella) [reply vsock/src/thread_backend.rs:59]: Agree

File: vsock/src/thread_backend.rs
Comment (stefano-garzarella) [reply vsock/src/thread_backend.rs:75]: Yep, I'll change

File: vsock/src/vhu_vsock.rs
Comment (stefano-garzarella) [reply vsock/src/vhu_vsock.rs:268]: Can we do this as a follow-up PR? (maybe opening an issue and when someone has time, we can fix)

File: vsock/src/main.rs
Comment (stefano-garzarella) [reply vsock/src/main.rs:20]: > About code coverage, I looked at kcov_output/ to see what all is missed and there are things we can easily cover I think. For example main.rs::start_backend_server() and a lot of other code in vhu_vsock_thread.rs which is not covered as of now in tests.

I agree, but can we postpone?
I already spent a lot of time to rebase Harsha's work, so my plan will be to fix the simple things and open issues to postpone the rest. (maybe some student can start with tests or refactoring, or I can do it when I'll have more time)
Is that okay?

Comment (stefano-garzarella) [reply vsock/src/main.rs:20]: Ack

File: vsock/src/main.rs
Comment (vireshk) [reply vsock/src/main.rs:70]: That would be nice. Thanks.

File: vsock/src/main.rs
Comment (vireshk) [reply vsock/src/main.rs:33]: Maybe then we should convert existing backends to use this nomenclature. I just want them to be consistent and yes daemon name doesn't matter much otherwise.

File: vsock/src/thread_backend.rs
Comment (vireshk) [reply vsock/src/thread_backend.rs:112]: its okay then, no issues.

**Comment by vireshk** [comment]
> I agree, but can we postpone? I already spent a lot of time to rebase Harsha's work, so my plan will be to fix the simple things and open issues to postpone the rest. (maybe some student can start with tests or refactoring, or I can do it when I'll have more time) Is that okay?

It is always better / easier to fix things at review time as they are lost otherwise. But I do get what you are saying here.

Please do whatever changes you can to the current patches, based on current comments, and we will get it merged. I can also try to spend some cycles on this then and raise issues too.

**Comment by stefano-garzarella** [comment]
> It is always better / easier to fix things at review time as they are lost otherwise. But I do get what you are saying here.
> 

I totally agree with you, but since I'm not the author, it's like doing a review every time to see if what I'm testing or changing makes sense :-)

> Please do whatever changes you can to the current patches, based on current comments, and we will get it merged. I can also try to spend some cycles on this then and raise issues too.

Thank you for understanding, I will do my best.
If you need help maintaining this workspace, I can allocate some time to help and add myself to `CODEOWNERS`.

File: vsock/src/main.rs
Comment (vireshk) [reply vsock/src/main.rs:33]: Maybe just keep vhost-user-vsock for now ? We can discuss later in a separate pull request on how to name all of these.

Comment (stefano-garzarella) [reply vsock/src/main.rs:33]: Ack

File: vsock/src/thread_backend.rs
Comment (stefano-garzarella) [reply vsock/src/thread_backend.rs:75]: I just realized that the value returned by `pop_front()` is `std::option::Option` and `.map_err` is not available.
Is there any other nice way? otherwise I'll leave it like this.

Comment (vireshk) [reply vsock/src/thread_backend.rs:75]: Maybe somthing like this ?

let key = self.backend_rxq.pop_front().ok_or(Error::EmptyBackendRxQ)?;

File: vsock/src/main.rs
Comment (stefano-garzarella) [reply vsock/src/main.rs:70]: `VsockConfig` is also used by `VhostUserVsockBackend` defined in vhu_vsock.rs
I think the original idea was to export part of the crate also as a library (in the future it will be nice, since I have a couple of use cases).
So for now I'll move only VsockArgs in main.rs, since it is really only needed by the binary.

File: vsock/src/vhu_vsock.rs
Comment (stefano-garzarella) [reply vsock/src/vhu_vsock.rs:268]: Fixed, it was just in the tests.

File: vsock/src/vhu_vsock.rs
Comment (stefano-garzarella) [reply vsock/src/vhu_vsock.rs:3]: I don't know, I'll remove

File: vsock/src/main.rs
Comment (stefano-garzarella) [reply vsock/src/main.rs:20]: In i2c and gpio I see `use log...` in the first group. IIUC `log` is an external crate. Should it be in the second group?

File: vsock/src/vhu_vsock.rs
Comment (stefano-garzarella) [reply vsock/src/vhu_vsock.rs:234]: yep

File: vsock/src/vhu_vsock.rs
Comment (stefano-garzarella) [reply vsock/src/vhu_vsock.rs:286]: I think so, I'll remove

File: vsock/src/vhu_vsock.rs
Comment (stefano-garzarella) [reply vsock/src/vhu_vsock.rs:214]: yep, make sense to me!

**Comment by stefano-garzarella** [comment]
v6:
 - addressed most of @vireshk requests
   - reordered `use` declarations
   - removed "vsock_" prefix where is redundant
   - moved VsockArgs in main.rs
   - used pub(crate) for types used only by our binary
   - and other small changes merged on previous commits

TODO:
- [ ] extend test coverage (I prefer as a follow up in a new PR, I'll open an issue to track)

**Comment by stefano-garzarella** [comment]
@vireshk @stsquad @mathieupoirier I just pressed the button to request Alex's review and GH messed up other requests. Sorry about that ;-)

**Comment by mathieupoirier** [comment]
I see that Viresh has beaten me to the punch on reviewing this pull request.  There is no need to do things twice, especially with a patchset this big, and as such will not continue with this work.

**Review by vireshk — APPROVED** [review]

**Comment by vireshk** [comment]
> I see that Viresh has beaten me to the punch on reviewing this pull request. There is no need to do things twice, especially with a patchset this big, and as such will not continue with this work.

You can approve it for now, I have already done so and we will continue to modify it to make it similar to other crates and fix any more issues we find later.

**Review by mathieupoirier — APPROVED** [review]

File: vsock/src/main.rs
Comment (harshanavkis) [reply vsock/src/main.rs:1]: Fine by me.

Comment (stefano-garzarella) [reply vsock/src/main.rs:1]: @harshanavkis thanks for your ACK, I'll prepare a PR to re-license vhost-user-vsock as Apache-2.0 only.

**Pull request merged**
